File: python/sampleONNXYolov5/yolov5.py
#!/usr/bin/env python3
# =======================================================================
# Copyright 2020-2023 Enflame. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# =======================================================================
#

import time
import argparse
import os
import glob
import cv2
import numpy as np
import TopsInference
import coco_labels


def model_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--device", type=str, default="dtu", help="Which device to run inference"
    )
    parser.add_argument(
        "--card_id",
        type=int,
        default=0,
        help="On which card to run inference processing",
    )
    parser.add_argument(
        "--cluster_ids",
        nargs="+",
        type=int,
        default=0,
        help="On which clusters to run inference. Generally now, \
              cluster_ids is a list with max length 6, it can be \
              [0], [0, 1], [0, 1, 2, 3, 4, 5] or any other range from 0 to 4. \
              cluster_ids can also be conveniently set to -1 to delegate \
              [0, 1, 2, 3, 4, 5]",
    )
    parser.add_argument(
        "--input_names",
        default="images",
        help="Input tensor names, which must be consistent with the content \
        of model file When there are multi input, names are seperated by \
        a comma.",
    )
    parser.add_argument(
        "--input_shape",
        type=str,
        default="1, 3, 640, 640",
        help="The standrd input shape of onnx model",
    )
    parser.add_argument("--data_format", type=str, default="NCHW")
    parser.add_argument(
        "--model_path",
        type=str,
        default="yolov5s.onnx",
        help="Which model to run inference,\
                        should be one of yolov5s, yolov5m, yolov5n, yolov5l, \
                        yolov5x.",
    )
    parser.add_argument("--engine", type=str, help="engine file to be reused.")
    parser.add_argument(
        "--data_path", type=str, help="path to image file or directory(only jpeg now)"
    )
    parser.add_argument(
        "--output_names",
        default="output",
        help="Output tensor names, which must be consistent with the content \
        of model file When there are multi input, names are seperated by \
        a comma.",
    )
    parser.add_argument(
        "--score_threshold",
        type=float,
        default=0.25,
    )
    parser.add_argument(
        "--iou_threshold", type=float, default=0.45, help="The IOU threshold during NMS"
    )
    parser.add_argument(
        "--display", action="store_true", help="Display or save the result."
    )
    parser.add_argument("--result_dir", type=str, default="./")

    arg_pars = parser.parse_args()
    return arg_pars


class Yolov5:
    time_preprocess = None
    time_infer = None
    time_postprocess = None
    time_total = None

    def __init__(self, arguments, onnx_file, device="dtu"):
        self.device = device
        # dtu device
        self.handler = TopsInference.set_device(
            arguments.card_id, arguments.cluster_ids
        )
        # parse onnx
        parser = TopsInference.create_parser(TopsInference.ONNX_MODEL)
        parser.set_input_names(arguments.input_names)
        parser.set_output_names(arguments.output_names)
        parser.set_input_shapes(arguments.input_shape)
        parser.set_input_dtypes("DT_FLOAT32")

        if arguments.engine is None:
            network = parser.read(onnx_file)
            optimizer = TopsInference.create_optimizer()
            # optimizer.set_build_flag(TopsInference.KFP16)
            # optimizer.set_build_flag(TopsInference.KDEFAULT)
            optimizer.set_build_flag(TopsInference.KFP16_MIX)
            self.engine = optimizer.build(network)
            engine_path = os.path.join(".", os.path.basename(onnx_file) + ".exec")
            self.engine.save_executable(engine_path)
        else:
            self.engine = TopsInference.PyEngine()
            self.engine.load_tops_executable(arguments.engine)

        assert self.engine is not None

        self.normH = int(arguments.input_shape.split(",")[2])
        self.normW = int(arguments.input_shape.split(",")[-1])
        self.img_scale = None
        # thresholds
        self.score_th = arguments.score_threshold
        self.nms_th = arguments.iou_threshold

    def __call__(self, images):
        inputs = []
        outputs = []

        time_start = time.time()

        for image in images:
            processed_img = self.preprocess(image)
            processed_img = np.around(processed_img, decimals=5)
            inputs.append(processed_img)

        time_after_preprocess = time.time()

        # Use sync mode, py_stream=None by default
        py_future = self.engine.run_with_batch(
            sample_nums=len(images),
            input_list=np.array([inputs]),
            buffer_type=TopsInference.TIF_ENGINE_RSC_IN_HOST_OUT_HOST,
        )

        time_after_infer = time.time()

        pred = py_future.get()
        pred = np.asarray(pred)
        outputs.append(pred)
        outputs = outputs[0][0]  # 1,1,25200,85

        # postprocessing
        results = []
        for index, image in enumerate(images):
            image_shape = image.shape
            result = self.postprocess(outputs[index], image_shape[1], image_shape[0])
            results.append(result)

        time_finish = time.time()
        self.time_preprocess = (time_after_preprocess - time_start) * 1000
        self.time_infer = (time_after_infer - time_after_preprocess) * 1000
        self.time_postprocess = (time_finish - time_after_infer) * 1000
        self.time_total = (time_finish - time_start) * 1000
        return results

    def destroy(self):
        TopsInference.release_device(self.handler)

    def nms(self, boxes, scores):
        if scores.shape[0] < 1 or boxes.shape[0] != scores.shape[0]:
            return None
        order = scores.argsort()[::-1]
        x1 = boxes[:, 0]
        y1 = boxes[:, 1]
        x2 = boxes[:, 2]
        y2 = boxes[:, 3]
        area = (y2 - y1 + 1) * (x2 - x1 + 1)
        boxes_left = []
        while len(order) > 0:
            i = order[0]
            boxes_left.append(i)
            xx1 = np.maximum(x1[i], x1[order[1:]])
            yy1 = np.maximum(y1[i], y1[order[1:]])
            xx2 = np.minimum(x2[i], x2[order[1:]])
            yy2 = np.minimum(y2[i], y2[order[1:]])
            w = np.maximum(0.0, xx2 - xx1 + 1)
            h = np.maximum(0.0, yy2 - yy1 + 1)
            inter = w * h
            iou = inter / (area[i] + area[order[1:]] - inter)
            ids = np.where(iou <= self.nms_th)[0]
            order = order[ids + 1]
        return boxes_left

    def preprocess(self, img):
        assert isinstance(
            img, np.ndarray
        ), "input image data type should be numpy.ndarray"
        assert img.shape[0] > 0 and img.shape[1] > 0, "input image shape error"

        shape = img.shape[0:2]
        self.im_scale = min(self.normH / img.shape[0], self.normW / img.shape[1])
        new_imsize = int(round(shape[1] * self.im_scale)), int(
            round(shape[0] * self.im_scale)
        )  # w, h
        dw, dh = self.normW - new_imsize[0], self.normH - new_imsize[1]  # wh padding
        dw /= 2
        dh /= 2
        if shape[::-1] != new_imsize:
            img = cv2.resize(img, new_imsize, interpolation=cv2.INTER_LINEAR).astype(
                np.float32
            )
        top, bottom = int(round(dh - 0.1)), int(round(dh + 0.1))
        left, right = int(round(dw - 0.1)), int(round(dw + 0.1))
        img = cv2.copyMakeBorder(
            img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=(114, 114, 114)
        )  # add border

        img = img.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB
        img = np.ascontiguousarray(img).astype(np.float32)
        img /= 255.0
        if len(img.shape) == 3:
            img = img[None]
        return img

    def postprocess(self, buffer, img_width, img_height):
        cands = buffer[..., 4] > self.score_th
        output = []
        scale = min(self.normW / img_width, self.normH / img_height)
        pad = (self.normW - img_width * scale) / 2, (
            self.normH - img_height * scale
        ) / 2
        # for x in buffer:
        x = buffer[cands]
        if not x.shape[0]:
            return []
        x[:, 5:] *= x[:, 4:5]
        boxes = xywh2xyxy(x[:, :4])
        conf = x[:, 4]
        cls = np.argmax(x[:, 5:], axis=1)
        box_num = boxes.shape[0]
        x = np.concatenate(
            (boxes, np.reshape(conf, (box_num, -1)), np.reshape(cls, (box_num, -1))),
            axis=1,
        )
        if x.shape[0] == 1:
            output.append(x)
        else:
            selected_boxes = self.nms(x[:, :4], x[:, 4])
            for box_id in selected_boxes:
                output.append(x[box_id])
        for i, det in enumerate(output):
            if min(det.shape) != 0:
                det[[0, 2]] -= pad[0]
                det[[1, 3]] -= pad[1]
                det[:4] /= scale
                det[[0, 2]].clip(0, img_height)
                det[[1, 3]].clip(0, img_width)
                # det[2].clamp_(0, img_height)
                # det[3].clamp_(0, img_width)
        return output


def display_results(image, results):
    display = image.copy()
    for box in results:
        x1, y1, x2, y2, cls = box[0], box[1], box[2], box[3], box[5]
        conf = str(np.around(box[4].item() * 100, decimals=2)) + "%"

        cv2.rectangle(
            display, (int(x1), int(y1)), (int(x2), int(y2)), (0, 0, 255), thickness=2
        )
        label = coco_labels.COCOLabels(int(cls)).name + ":" + conf
        label_size = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 1)[0]
        y1 = max(y1, label_size[1])
        cv2.putText(
            display,
            label,
            (int(x1), int(y1)),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            thickness=2,
        )
    return display


def xywh2xyxy(x):
    y = np.copy(x)
    y[:, 0] = x[:, 0] - x[:, 2] / 2  # top left x
    y[:, 1] = x[:, 1] - x[:, 3] / 2  # top left y
    y[:, 2] = x[:, 0] + x[:, 2] / 2  # bottom right x
    y[:, 3] = x[:, 1] + x[:, 3] / 2  # bottom right y
    return y


def main(arguments):
    images = []

    if os.path.isfile(arguments.data_path):
        image = cv2.imread(arguments.data_path)
        assert len(image.shape) > 0, "no image read"

        images.append(image)

    # Only jpg
    if os.path.isdir(arguments.data_path):
        jpg_list = glob.glob(arguments.data_path + "/*.jpg")

        for image in jpg_list:
            image = cv2.imread(os.path.join(arguments.data_path, image))
            assert len(image.shape) > 0, "no image read"

            images.append(image)

    model_file = arguments.model_path
    detector = Yolov5(arguments, onnx_file=model_file, device=arguments.device)

    # warmup
    for i in range(0, 3):
        results = detector(images)

    results = detector(images)

    time_list = []
    time_list.append(
        [
            detector.time_preprocess,
            detector.time_infer,
            detector.time_postprocess,
            detector.time_total,
        ]
    )

    # display or save
    for index, image in enumerate(images):
        if len(results) != 0:
            display = display_results(image, results[index])
        else:
            display = image

        img_name = str(index) + ".jpg"

        if not os.path.exists(arguments.result_dir):
            os.mkdir(arguments.result_dir)

        if isinstance(display, np.ndarray):
            if not arguments.display:
                cv2.imwrite(os.path.join(arguments.result_dir, img_name), display)
            else:
                cv2.imshow("yolov5 results", display)
                cv2.waitKey(0)

    print("{}{}time cost{}".format("\n" * 10, "*" * 40, "*" * 40))
    print(
        "time_preprocess,{} time_infer,{} time_postprocess,{} time_total".format(
            " " * 3, " " * 10, " " * 5
        )
    )
    for t in time_list:
        print("{}".format(t))
    print("the result has been saved to ", arguments.result_dir)


if __name__ == "__main__":
    arguments = model_arguments()
    main(arguments)

File: cpp/sampleONNXYolov5/yolov5.cpp
/*=======================================================================
 * Copyright 2020-2023 Enflame. All Rights Reserved.
 *
 *Licensed under the Apache License, Version 2.0 (the "License");
 *you may not use this file except in compliance with the License.
 *You may obtain a copy of the License at
 *
 *http://www.apache.org/licenses/LICENSE-2.0
 *
 *Unless required by applicable law or agreed to in writing, software
 *distributed under the License is distributed on an "AS IS" BASIS,
 *WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *See the License for the specific language governing permissions and
 *limitations under the License.
 *=======================================================================
 */

#include <unistd.h>

#include <algorithm>
#include <cassert>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <iostream>
#include <map>
#include <vector>

#include "../utils/tops_utils.h"
#include "TopsInference/TopsInferRuntime.h"

std::map<int, std::string> label_map = {
    {0, "person"},         {1, "bicycle"},       {2, "car"},
    {3, "motorbike"},      {4, "aeroplane"},     {5, "bus"},
    {6, "train"},          {7, "truck"},         {8, "boat"},
    {9, "traffic_light"},  {10, "fire_hydrant"}, {11, "stop_sign"},
    {12, "parking_meter"}, {13, "bench"},        {14, "bird"},
    {15, "cat"},           {16, "dog"},          {17, "horse"},
    {18, "sheep"},         {19, "cow"},          {20, "elephant"},
    {21, "bear"},          {22, "zebra"},        {23, "giraffe"},
    {24, "backpack"},      {25, "umbrella"},     {26, "handbag"},
    {27, "tie"},           {28, "suitcase"},     {29, "frisbee"},
    {30, "skis"},          {31, "snowboard"},    {32, "sports_ball"},
    {33, "kite"},          {34, "baseball_bat"}, {35, "baseball_glove"},
    {36, "skateboard"},    {37, "surfboard"},    {38, "tennis_racket"},
    {39, "bottle"},        {40, "wine_glass"},   {41, "cup"},
    {42, "fork"},          {43, "knife"},        {44, "spoon"},
    {45, "bowl"},          {46, "banana"},       {47, "apple"},
    {48, "sandwich"},      {49, "orange"},       {50, "broccoli"},
    {51, "carrot"},        {52, "hot_dog"},      {53, "pizza"},
    {54, "donut"},         {55, "cake"},         {56, "chair"},
    {57, "sofa"},          {58, "pottedplant"},  {59, "bed"},
    {60, "diningtable"},   {61, "toilet"},       {62, "tvmonitor"},
    {63, "laptop"},        {64, "mouse"},        {65, "remote"},
    {66, "keyboard"},      {67, "cell_phone"},   {68, "microwave"},
    {69, "oven"},          {70, "toaster"},      {71, "sink"},
    {72, "refrigerator"},  {73, "book"},         {74, "clock"},
    {75, "vase"},          {76, "scissors"},     {77, "teddy_bear"},
    {78, "hair_drier"},    {79, "toothbrush"},
};

typedef struct Point {
  float x;
  float y;
} Point;

typedef struct Rect {
  float left;
  float top;
  float width;
  float height;

  Rect(float l, float t, float w, float h)
      : left(l), top(t), width(w), height(h) {}

  void dump() {
    std::cout << "(" << left << "," << top << "," << width << "," << height
              << ")" << std::endl;
  }
} Rect;

char *getCmdOption(char **begin, char **end, const std::string &option) {
  char **itr = std::find(begin, end, option);
  if (itr != end && ++itr != end) {
    return *itr;
  }
  return 0;
}

bool cmdOptionExists(char **begin, char **end, const std::string &option) {
  return std::find(begin, end, option) != end;
}

void clamp(float &val, const float low, const float high) {
  if (val > high) {
    val = high;
  } else if (val < low) {
    val = low;
  }
}

static void sort(int n, const std::vector<float> x, std::vector<int> indices) {
  int i, j;
  for (i = 0; i < n; i++)
    for (j = i + 1; j < n; j++) {
      if (x[indices[j]] > x[indices[i]]) {
        // float x_tmp = x[i];
        int index_tmp = indices[i];
        // x[i] = x[j];
        indices[i] = indices[j];
        // x[j] = x_tmp;
        indices[j] = index_tmp;
      }
    }
}

bool nonMaximumSuppression(const std::vector<Rect> rects,
                           const std::vector<float> score,
                           float overlap_threshold,
                           std::vector<int> &index_out) {
  int num_boxes = rects.size();
  int i, j;
  std::vector<float> box_area(num_boxes);
  std::vector<int> indices(num_boxes);
  std::vector<int> is_suppressed(num_boxes);

  for (i = 0; i < num_boxes; i++) {
    indices[i] = i;
    is_suppressed[i] = 0;
    box_area[i] = (float)((rects[i].width + 1) * (rects[i].height + 1));
  }

  sort(num_boxes, score, indices);

  for (i = 0; i < num_boxes; i++) {
    if (!is_suppressed[indices[i]]) {
      for (j = i + 1; j < num_boxes; j++) {
        if (!is_suppressed[indices[j]]) {
          int x1max = std::max(rects[indices[i]].left, rects[indices[j]].left);
          int x2min =
              std::min(rects[indices[i]].left + rects[indices[i]].width,
                       rects[indices[j]].left + rects[indices[j]].width);
          int y1max = std::max(rects[indices[i]].top, rects[indices[j]].top);
          int y2min =
              std::min(rects[indices[i]].top + rects[indices[i]].height,
                       rects[indices[j]].top + rects[indices[j]].height);
          int overlap_w = x2min - x1max + 1;
          int overlap_h = y2min - y1max + 1;
          if (overlap_w > 0 && overlap_h > 0) {
            float iou = (overlap_w * overlap_h) /
                        (box_area[indices[j]] + box_area[indices[i]] -
                         overlap_w * overlap_h);
            if (iou > overlap_threshold) {
              is_suppressed[indices[j]] = 1;
            }
          }
        }
      }
    }
  }

  for (i = 0; i < num_boxes; i++) {
    if (!is_suppressed[i])
      index_out.push_back(i);
  }

  return true;
}

std::vector<Rect> postProcess(std::vector<float> &pred, const float img_height,
                              const float img_width, size_t bbox_num,
                              size_t bbox_len, const float conf_thres,
                              const float iou_thres) {
  std::vector<std::vector<float>> bboxes;
  for (size_t i = 0; i < bbox_num; i++) {
    std::vector<float> det;
    for (size_t j = 0; j < bbox_len; j++) {
      det.push_back(pred[i * 85 + j]);
    }
    bboxes.push_back(det);
  }

  // FIXME: scale is always 1.0 in this ppm case
  float scale = std::min(640.0 / img_width, 640.0 / img_height);
  std::vector<Rect> selected_boxes;
  std::vector<float> confidence;
  std::vector<float> class_id;

  // conf filter, calculate cls conf
  for (std::vector<float> box : bboxes) {
    float conf = box[4];
    if (conf > conf_thres) {
      confidence.push_back(conf);

      float max_conf = 0.0;
      int id = 0;

      for (size_t i = 5; i < box.size(); i++) {
        box[i] *= conf;
        if (box[i] > max_conf) {
          max_conf = box[i];
          id = i - 5;
        }
      }

      class_id.push_back(id);
      float centerX = box[0];
      float centerY = box[1];
      float width = box[2];
      float height = box[3];
      float left = centerX - width / 2.0;
      float top = centerY - height / 2.0;
      selected_boxes.push_back(Rect(left, top, width, height));
    }
  }

  // no object at all
  if (selected_boxes.size() == 0) {
    std::cout << "no bbox over score threshold detected." << std::endl;
    return {};
  }

  std::vector<Rect> det_result;
  std::vector<Rect> nms_out_rects;
  std::vector<int> indexes;
  nonMaximumSuppression(selected_boxes, confidence, iou_thres, indexes);

  for (int id : indexes) {
    det_result.push_back(selected_boxes[id]);
    std::cout << "cls: " << label_map[class_id[id]]
              << " conf: " << confidence[id] << " (" << selected_boxes[id].left
              << "," << selected_boxes[id].top << ","
              << selected_boxes[id].width << "," << selected_boxes[id].height
              << ")" << std::endl;
  }

  for (Rect &result_box : det_result) {
    result_box.left = (result_box.left) / scale;
    result_box.top = (result_box.top) / scale;
    result_box.width = (result_box.width) / scale;
    result_box.height = (result_box.height) / scale;

    clamp(result_box.left, 0, img_width);
    clamp(result_box.top, 0, img_height);
    clamp(result_box.width, 0, img_width);
    clamp(result_box.height, 0, img_height);
  }

  return det_result;
}

void writePPMFileWithBBox(const std::string &filename, uint8_t *ppm,
                          const std::vector<Rect> &bboxs, int ppm_h, int ppm_w,
                          int ppm_max) {
  std::ofstream outfile("./" + filename, std::ofstream::binary);
  assert(!outfile.fail());

  outfile << "P6"
          << "\n"
          << ppm_w << " " << ppm_h << "\n"
          << ppm_max << "\n";

  auto round = [](float x) -> int { return int(std::floor(x + 0.5f)); };

  for (auto &bbox : bboxs) {
    const int x1 = std::min(std::max(0, round(int(bbox.left))), ppm_w - 1);
    const int x2 =
        std::min(std::max(0, round(int(bbox.left + bbox.width))), ppm_w - 1);
    const int y1 = std::min(std::max(0, round(int(bbox.top))), ppm_h - 1);
    const int y2 =
        std::min(std::max(0, round(int(bbox.top + bbox.height))), ppm_h - 1);

    for (int x = x1; x <= x2; ++x) {
      // bbox top border
      ppm[(y1 * ppm_w + x) * 3] = 255;
      ppm[(y1 * ppm_w + x) * 3 + 1] = 0;
      ppm[(y1 * ppm_w + x) * 3 + 2] = 0;
      // bbox bottom border
      ppm[(y2 * ppm_w + x) * 3] = 255;
      ppm[(y2 * ppm_w + x) * 3 + 1] = 0;
      ppm[(y2 * ppm_w + x) * 3 + 2] = 0;
    }

    for (int y = y1; y <= y2; ++y) {
      // bbox left border
      ppm[(y * ppm_w + x1) * 3] = 255;
      ppm[(y * ppm_w + x1) * 3 + 1] = 0;
      ppm[(y * ppm_w + x1) * 3 + 2] = 0;
      // bbox right border
      ppm[(y * ppm_w + x2) * 3] = 255;
      ppm[(y * ppm_w + x2) * 3 + 1] = 0;
      ppm[(y * ppm_w + x2) * 3 + 2] = 0;
    }
  }

  outfile.write(reinterpret_cast<char *>(ppm), ppm_w * ppm_h * 3);
}

int main(int argc, char **argv) {
  if (cmdOptionExists(argv, argv + argc, "-h")) {
    std::cout << "usage: yolov5s [-h]" << std::endl;
    std::cout << "               [-o ONNX MODEL FILE PATH]" << std::endl;
    std::cout << "               [-i PPM IMAGE(640*640*3) FILE PATH]"
              << std::endl;
    std::cout << "               [-n NAME OF MODEL INPUT, DEFAULT IS: images]"
              << std::endl;
    std::cout << "               [-s INPUT SHAPE, DEFAULT IS: (1,3,640,640)]"
              << std::endl;
    std::cout << "               [-c SCORE THRESHOLD, DEFAULT IS: 0.25]"
              << std::endl;
    std::cout << "               [-u IOU THRESHOLD, DEFAULT IS: 0.45]"
              << std::endl;

    return 0;
  }

  const char *onnx_path;
  if (cmdOptionExists(argv, argv + argc, "-o")) {
    onnx_path = getCmdOption(argv, argv + argc, "-o");
  } else {
    // onnx_path = "../../models/yolov5s.onnx";
    std::cout << "[ERROR] Must specify ONNX model path" << std::endl;
    exit(-1);
  }

  const char *img_path;
  if (cmdOptionExists(argv, argv + argc, "-i")) {
    img_path = getCmdOption(argv, argv + argc, "-i");
  } else {
    img_path = "../../data/yolov5/zidane640.ppm";
  }

  const char *input_names;
  if (cmdOptionExists(argv, argv + argc, "-n")) {
    input_names = getCmdOption(argv, argv + argc, "-n");
  } else {
    input_names = "images";
  }

  const char *input_shapes;
  if (cmdOptionExists(argv, argv + argc, "-s")) {
    input_shapes = getCmdOption(argv, argv + argc, "-s");
  } else {
    input_shapes = "1,3,640,640";
  }

  float conf_thres = 0.25;
  float iou_thres = 0.45;

  char *conf_thres_char;
  if (cmdOptionExists(argv, argv + argc, "-c")) {
    conf_thres_char = getCmdOption(argv, argv + argc, "-c");
    conf_thres = atof(conf_thres_char);
  }

  char *iou_thres_char;
  if (cmdOptionExists(argv, argv + argc, "-u")) {
    iou_thres_char = getCmdOption(argv, argv + argc, "-u");
    iou_thres = atof(iou_thres_char);
  }

  int precision_type = 2; //

  // Note: Now only support inference on the preset 640 * 640 ppm
  int img_w = 640;
  int img_h = 640;

  std::string exec_path =
      engine_name_construct(onnx_path, "../../engines", atoi(input_shapes),
                            get_precision_str(precision_type));

  int card_id = 0;
  // uint32_t cluster_ids[] = {0, 1, 2, 3, 4, 5};  // use all pg to compute
  // uint32_t cluster_num = 6;                     // count of cluster_ids
  uint32_t cluster_ids[] = {0}; // use one pg to compute
  uint32_t cluster_num = 1;     // count of cluster_ids

  int batch_size = 1;

  // 1. init
  TopsInference::topsInference_init();
  void *tops_handler_ =
      TopsInference::set_device(card_id, cluster_ids, cluster_num);

  // 2. load engine file or build it from onnx file
  TopsInference::IEngine *engine = loadOrCreateEngine(
      exec_path.c_str(), onnx_path,
      TopsInference::BuildFlag::TIF_KTYPE_MIX_FP16, input_names, input_shapes);

  // 3. get input&output info from engine
  std::vector<ShapeInfo> inputs_shape_info = get_inputs_shape(engine);
  std::vector<ShapeInfo> outputs_shape_info = get_outputs_shape(engine);

  //  Debug only
  for (auto &s : inputs_shape_info) {
    s.dump();
  }

  for (auto &s : outputs_shape_info) {
    s.dump();
  }

  // 4. alloc host memory for outputs
  std::vector<void *> outputs =
      alloc_host_memory(outputs_shape_info, batch_size);

  // 5. prepare data
  std::string fname, label;
  std::ifstream ppm_file;

  ppm_file.open(img_path);

  int ppm_h, ppm_w, ppm_max;
  std::string magic;
  ppm_file >> magic >> ppm_w >> ppm_h >> ppm_max;
  std::cout << "magic: " << magic << ", w: " << ppm_w << ", h: " << ppm_h
            << ", max: " << ppm_max << std::endl;

  ppm_file.seekg(1, ppm_file.cur);

  std::vector<uint8_t> ppm_data;
  size_t ppm_size = ppm_w * ppm_h * 3;
  ppm_data.resize(ppm_size);

  uint8_t *p = static_cast<uint8_t *>(ppm_data.data());
  ppm_file.read(reinterpret_cast<char *>(p), ppm_size);

  // HWC->CHW
  std::vector<float> tensor;
  for (int channel = 0; channel < 3; ++channel) {
    for (size_t i = channel; i < ppm_size;) {
      tensor.push_back(p[i] / 255.0);
      i += 3;
    }
  }

  ppm_file.close();

  // 6. run
  std::vector<void *> inputs_list;
  inputs_list.push_back(tensor.data());

  // warmup
  for (int i = 0; i < 3; ++i) {
    auto ret = engine->run_with_batch(
        batch_size, inputs_list.data(), outputs.data(),
        TopsInference::BufferType::TIF_ENGINE_RSC_IN_HOST_OUT_HOST);
    if (!ret) {
      std::cout << "engine run_with_batch failed." << std::endl;
      exit(-1);
    }
  }

  std::chrono::steady_clock::time_point t1 = std::chrono::steady_clock::now();
  auto ret = engine->run_with_batch(
      batch_size, inputs_list.data(), outputs.data(),
      TopsInference::BufferType::TIF_ENGINE_RSC_IN_HOST_OUT_HOST);
  if (!ret) {
    std::cout << "engine run_with_batch failed." << std::endl;
    exit(-1);
  }

  std::chrono::steady_clock::time_point t2 = std::chrono::steady_clock::now();
  auto time_diff =
      std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
  std::cout << "-------------------- " << std::endl;
  std::cout << "gcu inference took time:" << (int)time_diff << " ms"
            << std::endl;
  std::cout << "-------------------- " << std::endl;

  // 7. postprocess output
  std::vector<float> pred_output;

  size_t datum_num = outputs_shape_info[0].mem_size /
                     outputs_shape_info[0].dtype_size; // 22500 * 85
  float *pred_p = static_cast<float *>(outputs[0]);
  for (size_t i = 0; i < datum_num; ++i) {
    pred_output.push_back(pred_p[i]);
  }

  size_t bbox_num = outputs_shape_info[0].dims[1];
  size_t bbox_len = outputs_shape_info[0].dims[2];
  auto result = postProcess(pred_output, img_h, img_w, bbox_num, bbox_len,
                            conf_thres, iou_thres);

  // for debug
  // std::cout << "size of result: " << result.size() << std::endl;
  //// dump
  // for (auto r : result) {
  // r.dump();
  //}

  // 8: Drawing and save
  writePPMFileWithBBox("./result.ppm", static_cast<uint8_t *>(ppm_data.data()),
                       result, ppm_h, ppm_w, ppm_max);

  // 9. free host outputs memory
  free_host_memory(outputs);

  // 10. release
  TopsInference::release_engine(engine);
  TopsInference::release_device(tops_handler_);
  TopsInference::topsInference_finish();

  return 0;
}
